File: remote.py
#!/usr/bin/env python

from serial import Serial
from time import sleep
from itertools import product, imap
from operator import itemgetter

COLS = 96
BIT_PER_BYTE = 8
ROWS = 7
PIXELS = set(product(xrange(COLS), xrange(ROWS)))
GET_X = itemgetter(0)
GET_Y = itemgetter(1)

class Remote(object):
    def __init__(self, serial_class=Serial, port='/dev/ttyUSB0'):
        self.port = serial_class(port, 19200, timeout=0.1)
        self.framebuf = [None] * ROWS * COLS
        sleep(2)
        self.clip = set()
        self.set_disp_offset(0)
        self.set_ops_offset(0)
        self.set_scroll_speed(0)
        for pixel in PIXELS:
			self.set_pixel(pixel, False)
        self.flush_pixels()

    def write_pixeldata(self, b1, b2, block_pixels):
        assert valid_block(b1)
        assert valid_block(b2)
        x1, y1 = b1
        x2, y2 = b2
        assert (x2 - x1) * (y2 - y1) == len(block_pixels)
        raw = ''.join((
            chr(0xC0 | (y1 << 3) | y2),
            chr((x1 << 4) | x2),
            block_pixels))
        self.send_command_with_1b_reply(raw)

    def get_pixel(self, pixel):
        return self.framebuf[pixel2fbindex(pixel)]

    def set_pixel(self, pixel, value):
        self.framebuf[pixel2fbindex(pixel)] = value
        block = pixel2block(pixel)
        self.clip.add(block)

    def set_ops_offset(self, offset):
        return self.send_command_with_1b_reply(chr(0x20 | offset))

    def set_disp_offset(self, offset):
        return self.send_command_with_1b_reply(chr(0x02) + chr(offset))

    def set_scroll_speed(self, speed):
        return self.send_command_with_1b_reply(chr(0x40 | speed))

    def get_disp_offset(self):
        return self.send_command_with_1b_reply(chr(0x03))

    def write_eeprom(self, payload):
        length = len(payload)
        count = self.send_command_with_1b_reply(chr(0x80 | (length >> 8)) + chr(length & 0xFF))
        for b in payload:
            result = self.send_command_with_1b_reply(b)
            assert result == ord(b)
        return count

    def send_command_with_1b_reply(self, cmd):
        self.port.write(cmd)
        while True:
            p = self.port.read()
            if p:
                return ord(p)

    def flush_pixels(self):
        clip = self.clip
        if not clip:
            return
        xs = set(imap(GET_X, clip))
        ys = set(imap(GET_Y, clip))
        left = min(xs)
        right = max(xs) + 1
        top = min(ys)
        bottom = max(ys) + 1
        width = right - left
        pixeldata = ''.join(
                self.get_pixeldata_byte((left + offset % width, top + offset / width))
                for offset in xrange((bottom - top) * width))
        self.write_pixeldata((left, top), (right, bottom), pixeldata)
        self.clip.clear()

    def get_pixeldata_byte(self, block):
        retval = 0
        bx, by = block
        index = pixel2fbindex((bx * BIT_PER_BYTE, by))
        for pos, bit in enumerate(self.framebuf[index:index + 8]):
            if bit:
                retval |= 0x80 >> pos
        return chr(retval)

def pixel2fbindex(pixel):
    col, row = pixel
    return row * COLS + col

def pixel2block(pixel):
    col, row = pixel
    return col / BIT_PER_BYTE, row

def valid_block(block):
    x, y = block
    return 0 <= x <= (COLS / BIT_PER_BYTE) and 0 <= y <= ROWS

if __name__ == '__main__':
    r = Remote()
    for y in xrange(ROWS):
        for x in xrange(COLS):
            r.set_pixel((x, y), True)
            r.flush_pixels()
            sleep(0.05)
            r.set_pixel((x, y), False)

File: remote-cpp/Remote.cpp
#include "Remote.h"
#include <string.h>
#include <unistd.h>

#define INVALID_CLIP 0xFF

#define PIXEL_TO_FB_INDEX(x, y) (((y) * COLS + x) / BIT_PER_BYTE)
#define PIXEL_TO_FB_BIT(x) (1 << (7 - (x % BIT_PER_BYTE)))

#define FB_INDEX_TO_CLIP_INDEX(fbi) ((fbi) / BIT_PER_BYTE)
#define FB_INDEX_TO_CLIP_BIT(x) PIXEL_TO_FB_BIT(x)

Remote::Remote(int fdrd, int fdwr) {
	m_fdrd = fdrd;
	m_fdwr = fdwr;
	for (uint8_t y = 0; y < ROWS; y++) {
		for (uint8_t x = 0; x < COLS; x++) {
			setPixel(x, y, false);
		}
	}
	flushPixels();
}

/*
 * framebuf:
 *
 *     x=0      x=8          x=88
 * y=0 00000000 00000000 ... 00000000
 * y=1 00000000 00000000 ... 00000000
 *     ...      ...      ... ...
 * y=6 00000000 00000000 ... 00000000
 *
 * clip:
 *
 *     y=0          y=1           y=2  
 *     x=0      x=64x=0      x=64 x=0
 *     00000000 00000000 00000000 00000000 ...
 */ 

void Remote::flushPixels() {
	uint8_t left = INVALID_CLIP, right = INVALID_CLIP, top = INVALID_CLIP, bottom = INVALID_CLIP;
	uint8_t idx = 0, bit = 0x80;
	uint8_t clip = m_clip[0];
	for (uint8_t y = 0; y < ROWS; y++) { 
		for (uint8_t x = 0; x < COLS / BIT_PER_BYTE; x++) {
			if (clip & bit) {
				if (top == INVALID_CLIP || top > y) top = y;
				if (bottom == INVALID_CLIP || bottom <= y) bottom = y + 1;
				if (left == INVALID_CLIP || left > x) left = x;
				if (right == INVALID_CLIP || right <= x) right = x + 1;

			}
			if (bit == 1) {
				bit = 0x80;
				clip = m_clip[++idx];
			} else bit >>= 1;
		}
	}
	if (left == INVALID_CLIP) return;

	const uint8_t width = right - left;
	const uint8_t height = bottom - top;
	const size_t buflen = 2 + height * width;
	uint8_t buf[buflen];

	buf[0] = 0xC0 | (top << 3) | bottom;
	buf[1] = (left << 4) | right;
	for (uint8_t y = 0; y < height; y++) {
		memcpy(buf + 2 + y * width, m_framebuf + left + (y + top) * COLS / BIT_PER_BYTE, width);
	}

	write(m_fdwr, buf, buflen);
	while (!read(m_fdrd, buf, 1));

	memset(m_clip, 0x00, sizeof(m_clip));
}

void Remote::setPixel(uint8_t x, uint8_t y, bool value) {
	const uint8_t idx = PIXEL_TO_FB_INDEX(x, y);
	const uint8_t old = m_framebuf[idx];
	const uint8_t mask = PIXEL_TO_FB_BIT(x);
	if (value) {
		if (mask & old) return;
		m_framebuf[idx] = old | mask;
	} else {
		if ((mask & old) == 0) return;
		m_framebuf[idx] = old & ~mask;
	}
	m_clip[FB_INDEX_TO_CLIP_INDEX(idx)] |= FB_INDEX_TO_CLIP_BIT(idx);
}

bool Remote::getPixel(uint8_t x, uint8_t y) const {
	return (m_framebuf[PIXEL_TO_FB_INDEX(x, y)] & PIXEL_TO_FB_BIT(x)) != 0;
}
